sa/SASCM.INI

```diff
@@ -2214,3 +2215,296 @@ DATE=6.14.2013
+0D0F=2,set_car_model_alpha %1d% alpha %2d%
+0D10=2,set_char_model_alpha %1d% alpha %2d%
+0D11=2,set_object_model_alpha %1d% alpha %2d%
+0D2D=8,get_local_time_year_to %1d% month_to %2d% day_of_week_to %3d% day_to %4d% hour_to %5d% minute_to %6d% second_to %7d% milliseconds_to %8d%
+0D2E=3,set_thread %1d% var %2d% value %3d%
+0D2F=3,%3d% = get_thread %1d% var %2d% // keep NewOpcodes order
+0D33=3,set_car %1d% door %2d% window_state %3d%
+0E1E=15,draw_texture_plus %1d% event %2d% pos %3d% %4d% size %5d% %6d% angle %7d% depth %8d% fix_aspect_ratio %9d% maskTrisCount %10d% maskTrisArray %11d% rgba %12d% %13d% %14d% %15d% 
+0E3C=2,get_texture_from_sprite %1d% store_to %2d%
+0E62=8,print %1s% event %2d% at %3d% %4d% scale %5d% %6d% fixAR %7d% style %8d%
+0E63=27,print %1s% event %2d% at %3d% %4d% scale %5d% %6d% fixAR %7d% style %8d% prop %9d% align %10d% wrap %11d% justify %12d% color %13d% %14d% %15d% %16d% outline %17d% shadow %18d% dropColor %19d% %20d% %21d% %22d% background %23d% backColor %24d% %25d% %26d% %27d%
+0D1E=4,quat_slerp %1d% to %2d% lambda %3d% result %4d%
+0D16=2,set_matrix_rotation_from_quat %1d% rotation_from_quat %2d%
+0D17=2,set_quat_from_matrix %1d% to_quat %2d%
+0D18=6,rotate_quat_on_axis %1d% axis_vector %2d% %3d% %4d% angle %5d% combine_op %6d%
+0D19=2,get_normalised_quat %1d% to_quat %2d%
+0D1A=3,multiply_quats %3d% = quat %1d% * quat %2d% // keep NewOpcodes order
+0D24=5,set_quat %1d% elements %2d% %3d% %4d% %5d%
+0D29=5,get_quat %1d% elements_to %2d% %3d% %4d% %5d%
+0D01=6,rotate_matrix %1d% on_axis %2d% %3d% %4d% angle %5d% combine_op %6d%
+0D02=2,%2d% = matrix %1d% x_angle // keep NewOpcodes order
+0D03=2,%2d% = matrix %1d% y_angle // keep NewOpcodes order
+0D04=2,%2d% = matrix %1d% z_angle // keep NewOpcodes order
+0E1F=4,ease %1d% mode %2d% way %3d% to %4d%
+0E27=5,get_angle_from_two_coords %1d% %2d% and %3d% %4d% to %5d%
+0E03=2,perlin_noise %1d% store_to %2d%
+0E29=6,perlin_noise %1d% frequency %2d% amplitude %3d% lacunarity %4d% persistence %5d% store_to %6d%
+0EF0=6,get_coord_from_angled_distance %1d% %2d% angle %3d% dist %4d% store_to %5d% %6d%
+0EF1=7,perlin_noise_fractal_2d %1d% %2d% frequency %3d% amplitude %4d% lacunarity %5d% persistence %6d% store_to %7d%
+0EF2=8,perlin_noise_fractal_3d %1d% %2d% %3d% frequency %4d% amplitude %5d% lacunarity %6d% persistence %7d% store_to %8d%
+0EF4=4,clamp_float %1d% min %2d% max %3d% store_to %4d%
+0EF7=4,clamp_int %1d% min %2d% max %3d% store_to %4d%
+0EB3=4,convert_direction_to_quat %1d% dir %2d% %3d% %4d%
+0EF3=4,lerp %1d% %2d% %3d% store_to %4d%
+0E21=1,get_audio_sfx_volume %1d%
+0E22=1,get_audio_radio_volume %1d%
+0E3B=2,get_audiostream_internal %1d% store_to %2d%
+0E28=4,write_struct %1d% offset %2d% size %3d% value %4d%
+0D4E=4,%4d% = read_struct %1d% offset %2d% size %3d% // keep newOpcodes ordering for SB
+0D4F=4,write_struct %1d% offset %2d% size %3d% value %4d% // newOpcodes version of 0E28, the original one doesn't work, this is a fall off
```

Comment:
@JuniorDjjr I added CLEO+ definitions after NewOpcodes. 

Can you please explain what is the idea behind having newOpcodes definitions inter-mixed with CLEO+ ones? Can I remove duplicated opcodes?

Reply:
It includes the exact same definitions from newOpcodes, so, you can just remove all newOpcodes opcodes (0D**) from CLEO+ definitions without any problem. They are just retrocompatibility.

The opcodes ordering are from source code order (separated by categories), so the opcodes seems very mixed, and during development I removed and added other more useful commands for other categories etc, so don't expect anything organized in terms of numbering.

Reply:
Ok, thanks.

@JuniorDjjr Can you clarify one more thing for me? newOpcodes have these definitions:
```
0D02=2,%2d% = matrix %1d% y_angle
0D03=2,%2d% = matrix %1d% x_angle
```
in CLEO+ those defined as
```
0D02=2,%2d% = matrix %1d% x_angle // keep NewOpcodes order
0D03=2,%2d% = matrix %1d% y_angle // keep NewOpcodes order
```

note swapped X and Y in descriptions. Which version is correct?

Reply:
This is interesting.
The only place where 0D02=y is in newOpcodes' sascm.ini. In his source code, 0D02=x (considering the name, I'm not sure reading the code)
```
// 0D02=2,%2d% = matrix %1d% x_angle
// 0D02: 1@ = matrix 0@ x_angle
// SCR: GET_MATRIX_X_ANGLE
OpcodeResult WINAPI Matrix::GetXAngle(CScriptThread* thread)
{
	OpcodeParams params(thread, 2);
	RwMatrix *matrix;
	params >> matrix;
	float x = matrix->right.x;
	float y = matrix->right.y;
	float z = matrix->right.z;
	float angle = CGeneral::GetATanOfXY(z, sqrt(x * x + y * y)) * 57.295776f - 90.0f;
	while(angle < 0.0)
		angle += 360.0;
	params << angle;
    return OR_CONTINUE;
}
```
opcodes.txt is `0D02: 1@ = matrix 0@ x_angle`.
So, I'm not 100% sure, but everything indicates that it was a sascm.ini typo.
For Y angle, he uses the same code with `matrix->up`. Someone with more experience in 3D mathematics can confirm this.